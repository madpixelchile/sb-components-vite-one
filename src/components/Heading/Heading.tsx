import React from "react";
import { StyledHeading1, StyledHeading2, StyledHeading3, StyledHeading4, StyledHeading5, StyledHeading6 } from "./StyledHeadings";
import { AnyProp, ChildrenProp, ClassNameProp, LabelProp } from "../../types/types";


type HeadingSizeProp = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export interface HeadingProps extends
ClassNameProp,
ChildrenProp,
LabelProp,
AnyProp
{
    variant?: HeadingSizeProp;
    size?: HeadingSizeProp;
}

export const Heading: React.FC<HeadingProps> = (
    {
        children = null,
        variant = 'h1',
        label = 'Example Heading',
        size = '',
        className = '',
        ...props
    }
) => {

    const HeadingComponents: { [key: string]: React.ElementType } = {
        h1: StyledHeading1,
        h2: StyledHeading2,
        h3: StyledHeading3,
        h4: StyledHeading4,
        h5: StyledHeading5,
        h6: StyledHeading6,
    };

    const SelectedHeading = HeadingComponents[variant];

    return (
        <>
            <>
                {
                    SelectedHeading && (
                        <SelectedHeading className={`${className} ${size}`} {...props}>
                            {children || label}
                        </SelectedHeading>
                    )
                }
            </>

        </>
    )
}

export default Heading;

