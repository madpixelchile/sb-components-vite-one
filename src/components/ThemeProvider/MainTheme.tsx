

import { ThemeProvider } from 'styled-components';
import { StyledThemeProvider } from './StyledThemeProvider';
import { darkMode, lightMode } from '../../styles/theme';

export interface MainThemeProps {
  children?: React.ReactNode;
  theme?: string | {},
  [key: string]: any;
}

export const MainTheme = ({ children = null, theme = 'light' }:MainThemeProps) => {

  let selectedTheme: string | object = lightMode;

  if(typeof theme !== 'object'){
    if(theme === 'light'){
      //No object - string light - default lightModeAdded
      selectedTheme = lightMode;
    }else if(theme === 'dark'){
      //No object - string dark -  default darkModeAdded
      selectedTheme = darkMode;
    }
  }else if(typeof theme === 'object'){
    if(Object.keys(theme).length === 0){ //!PENDING Compare with the final BASE required properties
      //Empty object - Default lightMode added
      selectedTheme = lightMode;
    }else{
      //Full object - CustomMode or theme added
      selectedTheme = theme;
    }
  }

  return (
    <>
        <ThemeProvider theme={ selectedTheme }>
            <StyledThemeProvider/>
            { children }
        </ThemeProvider>
    </>
  )
}

export default  MainTheme;
