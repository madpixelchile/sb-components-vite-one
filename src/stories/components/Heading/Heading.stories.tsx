import { Meta, StoryObj } from "@storybook/react";
import Heading from "../../../components/Heading/Heading";
import MainTheme from "../../../components/ThemeProvider/MainTheme";

const meta = {
    //Se recomienda generar un storybook a parte del proyecto final y con vite
    //El slash genera la visualización en carpetas y subniveles en la navegación principal
    title: 'Components/headings/Heading',
    //El componente a incorporar en el muestrario
    component: Heading,

    tags:[
        'autodocs'
    ],
    //Para centrar el componente en la mesa o cama (muestrario) de componente en pantalla
    parameters: {
        layout: 'centered'
    },
    //Para que en la documentación los tamaños salgan en formato radio y hacia el lado
    argTypes:{
        size: { control: 'select', options: ['','h2','h3','h4','h5','h6'] },
        variant: { control: 'select', options: ['','h1','h2','h3','h4','h5','h6'] },
    },
} satisfies Meta<typeof Heading>;

export default meta;

type Story = StoryObj<typeof meta>

//Ejemplo de componente con sus props específicas aplicadas
export const Heading1 = ({...props}:Story)=>(
    <MainTheme>
        <Heading {...props} >Esto es un título</Heading>
    </MainTheme>
)


