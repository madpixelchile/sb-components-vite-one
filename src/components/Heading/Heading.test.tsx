
import { render } from '@testing-library/react'; //screen screen.debug();
import { Heading } from './Heading';
import MainTheme from '../ThemeProvider/MainTheme';

describe('Testing <Heading/>', () => {

    test('Testing className & Children', () => {

        const { getByText, container } = render(<MainTheme><Heading className='h2'>Hola mundo</Heading></MainTheme>);
        const childrenElement = getByText('Hola mundo');
        expect(childrenElement.innerHTML).toBe('Hola mundo');

        const classProp = container.getElementsByClassName('h2');
        expect(classProp.length).toBe(1)

        expect(container).toMatchSnapshot();

    })

    test('Testing variant', () => {

        const { getByText, getByRole } = render(<MainTheme><Heading variant='h2'>Hola mundo</Heading></MainTheme>);
        const childrenElement = getByText('Hola mundo');
        expect(childrenElement.innerHTML).toBe('Hola mundo');

        expect(getByRole('heading',{level:2}).innerHTML).toBe('Hola mundo')

    })

    test('Testing label without children prop', () => {

        const { getByText, getByRole } = render(<MainTheme><Heading variant='h2' label={'Hola mundo'}></Heading></MainTheme>);
        const childrenElement = getByText('Hola mundo');
        expect(childrenElement.innerHTML).toBe('Hola mundo');

        expect(getByRole('heading',{level:2}).innerHTML).toBe('Hola mundo')

    })

})