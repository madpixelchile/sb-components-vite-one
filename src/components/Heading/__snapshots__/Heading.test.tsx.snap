// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Heading/> Testing className & Children 1`] = `
<div>
  <h1
    class="sc-bdnyFh bomFrv h2 "
  >
    Hola mundo
  </h1>
</div>
`;
